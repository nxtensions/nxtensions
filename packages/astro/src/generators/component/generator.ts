import { formatFiles, Tree } from '@nx/devkit';
import type { GeneratorOptions } from './schema';
import {
  addComponentFile,
  addStyleDependencies,
  normalizeOptions,
} from './utilities';

export async function componentGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
) {
  const options = normalizeOptions(tree, rawOptions);

  addComponentFile(tree, options);
  const styleTask = addStyleDependencies(tree, options);
  await formatFiles(tree);

  return styleTask;
}

export default componentGenerator;
