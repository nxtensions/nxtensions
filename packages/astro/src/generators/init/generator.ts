import { cypressInitGenerator } from '@nrwl/cypress';
import {
  addDependenciesToPackageJson,
  GeneratorCallback,
  Tree,
} from '@nrwl/devkit';
import { GeneratorOptions } from './schema';
import {
  addProjectGraphPlugin,
  updateGitignore,
  addVSCodeRecommendedExtensions,
} from './utilities';
import { astroVersion } from './versions';

export function initGenerator(
  tree: Tree,
  options: GeneratorOptions
): GeneratorCallback {
  addProjectGraphPlugin(tree);
  updateGitignore(tree);
  addVSCodeRecommendedExtensions(tree);

  const tasks: GeneratorCallback[] = [];
  if (options.addCypressTests !== false) {
    const cypressTask = cypressInitGenerator(tree);
    tasks.push(cypressTask);
  }

  const depsTask = addDependenciesToPackageJson(
    tree,
    {},
    { astro: astroVersion }
  );
  tasks.push(depsTask);

  return () => tasks.forEach((task) => task());
}

export default initGenerator;
