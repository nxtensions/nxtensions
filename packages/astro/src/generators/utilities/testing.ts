import { Tree, readJsonFile, updateJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace as _createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { PackageJson } from '../../utilities/package-json';

export function createTreeWithEmptyWorkspace(nxVersion?: string): Tree {
  const tree = _createTreeWithEmptyWorkspace();
  if (!nxVersion) {
    const { devDependencies } = readJsonFile<PackageJson>('package.json');
    nxVersion = devDependencies['nx'];
  }
  updateJson(tree, 'package.json', (json) => {
    json.devDependencies['nx'] = nxVersion;
    return json;
  });

  return tree;
}
