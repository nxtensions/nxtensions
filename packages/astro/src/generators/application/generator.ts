import {
  ensurePackage,
  formatFiles,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { addDependenciesToPackageJson } from '../../utilities/package-json';
import { initGenerator } from '../init/generator';
import {
  astroCheckVersion,
  astroVersion,
  getInstalledNxVersion,
  typescriptVersion,
} from '../utilities/versions';
import type { GeneratorOptions } from './schema';
import {
  addFiles,
  addIntegrationsPackages,
  addProject,
  configureTailwindIntegration,
  normalizeOptions,
  setDefaultProject,
  setupE2ETests,
} from './utilities';

export async function applicationGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  return await applicationGeneratorInternal(tree, {
    projectNameAndRootFormat: 'derived',
    ...rawOptions,
  });
}

export async function applicationGeneratorInternal(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  const options = await normalizeOptions(tree, rawOptions);

  const tasks: GeneratorCallback[] = [];
  const { initGenerator: jsInitGenerator } = ensurePackage<
    typeof import('@nx/js')
  >('@nx/js', getInstalledNxVersion(tree));
  const jsTask = await jsInitGenerator(tree, { skipFormat: true });
  tasks.push(jsTask);

  await initGenerator(tree, {});

  const depsTask = addDependenciesToPackageJson(
    tree,
    {},
    {
      astro: astroVersion,
      '@astrojs/check': astroCheckVersion,
      typescript: typescriptVersion,
    }
  );
  tasks.push(depsTask);

  addProject(tree, options);
  setDefaultProject(tree, options.projectName);
  addFiles(tree, options);
  await configureTailwindIntegration(tree, options);

  const e2eTask = await setupE2ETests(tree, options);
  if (e2eTask) {
    tasks.push(e2eTask);
  }

  const integrationsTask = addIntegrationsPackages(tree, options.integrations);
  if (integrationsTask) {
    tasks.push(integrationsTask);
  }

  await formatFiles(tree);

  return () => tasks.forEach((task) => task());
}

export default applicationGenerator;
