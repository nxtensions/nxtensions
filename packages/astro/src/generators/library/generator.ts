import { formatFiles, GeneratorCallback, Tree } from '@nrwl/devkit';
import { initGenerator } from '../init/generator';
import { GeneratorOptions } from './schema';
import {
  addFiles,
  addPathMapping,
  addProject,
  normalizeOptions,
} from './utilities';

export async function libraryGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback> {
  if (rawOptions.publishable && !rawOptions.importPath) {
    throw new Error(
      'For publishable libraries the "--importPath" must be provided. Please note it needs to be a valid npm package name (e.g. my-lib or @my-org/my-lib).'
    );
  }

  const options = normalizeOptions(tree, rawOptions);

  const initTask = initGenerator(tree, { addCypressTests: false });

  addProject(tree, options);
  addFiles(tree, options);
  addPathMapping(tree, options);

  await formatFiles(tree);

  return initTask;
}

export default libraryGenerator;
