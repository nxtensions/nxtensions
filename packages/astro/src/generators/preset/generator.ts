import {
  formatFiles,
  joinPathFragments,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { applicationGenerator } from '../application/generator';
import type { GeneratorOptions } from './schema';

export default async function (
  tree: Tree,
  options: GeneratorOptions
): Promise<GeneratorCallback | void> {
  const applicationTask = await applicationGenerator(tree, {
    name: options.astroAppName,
    tags: options.tags,
    directory: joinPathFragments('apps', options.astroAppName),
    projectNameAndRootFormat: 'as-provided',
  });

  await formatFiles(tree);

  return applicationTask;
}
