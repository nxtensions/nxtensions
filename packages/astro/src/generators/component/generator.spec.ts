jest.mock('@nx/devkit', () => ({
  ...jest.requireActual('@nx/devkit'),
  formatFiles: jest.fn(),
}));

import {
  addProjectConfiguration,
  formatFiles,
  readJson,
  type Tree,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { componentGenerator } from './generator';
import type { Style } from './schema';

describe('component generator', () => {
  let tree: Tree;
  const appProject = 'app1';
  const libProject = 'lib1';

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();

    addProjectConfiguration(tree, appProject, {
      root: `apps/${appProject}`,
      projectType: 'application',
    });
    addProjectConfiguration(tree, libProject, {
      root: `libs/${libProject}`,
      projectType: 'library',
    });

    jest.clearAllMocks();
  });

  test('should create the component in the right location when adding it to an application', async () => {
    await componentGenerator(tree, {
      path: `apps/${appProject}/src/components/Foo`,
    });

    expect(
      tree.exists(`apps/${appProject}/src/components/Foo.astro`)
    ).toBeTruthy();
  });

  test('should create the component in the right location when adding it to a library', async () => {
    await componentGenerator(tree, {
      path: `libs/${libProject}/src/Foo`,
    });

    expect(tree.exists(`libs/${libProject}/src/Foo.astro`)).toBeTruthy();
  });

  test('should create the component in src/lib when that dir exists and it is not empty for backward compat', async () => {
    tree.write(`libs/${libProject}/src/lib/Cmp1.astro`, '<div>cmp1</div>');

    await componentGenerator(tree, {
      path: `libs/${libProject}/src/lib/Foo`,
    });

    expect(tree.exists(`libs/${libProject}/src/lib/Foo.astro`)).toBeTruthy();
  });

  test.each([['foo'], ['Foo'], ['foo-bar'], ['fooBar'], ['FooBar']])(
    'should create the component with "%s" as provided',
    async (name) => {
      await componentGenerator(tree, {
        path: `libs/${libProject}/src/${name}`,
      });

      expect(tree.exists(`libs/${libProject}/src/${name}.astro`)).toBeTruthy();
    }
  );

  test('should format files', async () => {
    await componentGenerator(tree, {
      path: `apps/${appProject}/src/components/foo`,
    });

    expect(formatFiles).toHaveBeenCalled();
  });

  describe('--styles', () => {
    test.each([undefined, 'css', 'scss', 'sass', 'less', 'styl', 'none'])(
      'should create the component with the right stylesheet content when "%s" is provided',
      async (style: Style) => {
        await componentGenerator(tree, {
          path: `apps/${appProject}/src/components/Foo`,
          style,
        });

        expect(
          tree.read(`apps/${appProject}/src/components/Foo.astro`, 'utf-8')
        ).toMatchSnapshot();
      }
    );

    test.each([
      ['scss', 'sass'],
      ['sass', 'sass'],
      ['less', 'less'],
      ['styl', 'stylus'],
    ])(
      'should install the right package when "%s" is provided',
      async (style: Style, packageName: string) => {
        await componentGenerator(tree, {
          path: `apps/${appProject}/src/components/foo`,
          style,
        });

        const { devDependencies } = readJson(tree, 'package.json');
        expect(devDependencies[packageName]).toBeTruthy();
      }
    );

    test.each([undefined, 'css', 'none'])(
      'should not install any package when "%s" is provided',
      async (style: Style) => {
        await componentGenerator(tree, {
          path: `apps/${appProject}/src/components/foo`,
          style,
        });

        const { devDependencies } = readJson(tree, 'package.json');
        expect(devDependencies).toStrictEqual({});
      }
    );
  });
});
