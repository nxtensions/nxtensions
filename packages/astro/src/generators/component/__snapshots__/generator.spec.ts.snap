// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component generator --styles should create the component with the right stylesheet content when "css" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style>
p {
  color: #ff5c01;
}
</style>
"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "less" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style lang="less">
p {
  color: #ff5c01;
}
</style>
"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "none" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>


"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "sass" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style lang="sass">
p
  color: #ff5c01
</style>
"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "scss" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style lang="scss">
p {
  color: #ff5c01;
}
</style>
"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "styl" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style lang="styl">
p {
  color: #ff5c01;
}
</style>
"
`;

exports[`component generator --styles should create the component with the right stylesheet content when "undefined" is provided 1`] = `
"---
// You can import other components here if needed.
// Feel free to delete this Frontmatter script if not needed.
---

<p>Foo works!</p>

<style>
p {
  color: #ff5c01;
}
</style>
"
`;
